### ISSUE TITLE
click/touch events appear fundamentally broken on mobile

### ISSUE THREAD

**Issue by chschie** [mapbox/mapbox.js#733]
A default, bare bones implementation of MapBox appears to 'eat' about every other click on android, and intermittently on iOS, to the point it would be impossible to use this system in any kind of production environment on mobile.

By 'eat clicks' I am speaking specifically about tapping the zoom or layer buttons, etc. Is the MapBox team aware of this issue and is there any time window to expect a fix?

**Comment by tmcw** [comment]
Can you rephrase this ticket in more concrete terms? What are you trying to do, what do you expect mapbox.js to do, and what does it do instead?

**Comment by mourner** [comment]
I think this is about click not firing on markers if you moved the finger a little, there was a patch for this https://github.com/Leaflet/Leaflet/pull/2503 — need to backport it to Leaflet 0.7.3

**Comment by chschie** [comment]
When I click on various items in the MapBox UI in a mobile environment, such as zoom, etc, I would expect that each touch/click would be registered.

However, it appears that every touch/click isn't registered on mobile, and that the user can often touch/click on zoom, for example, and nothing will happen, requiring the user to click a single UI element multiple times before it registered.

On a vanilla mapbox install, we are seeing this behavior out of the box on mobile.

**Comment by chschie** [comment]
To be clear, this issue has nothing to do with markers, and is about the default mapbox UI elements (ie, zoom, layer select, etc).

**Comment by mourner** [comment]
Can you specify exact versions of OS / browsers where this is reproducible?

**Comment by chschie** [comment]
OS: Android 4.3 (Samsung Galaxy S4)
Browser: Default Browser (ie, 'browser'), and Chrome v.34

OS: iOS 5.1.1 (Gen 1 iPad)
Browser: Safari

OS: iOS 5.1.1 (Gen 1 iPod touch)
Browser: Safari

OS: iPad 7.1 (iPad Air, newest version)
Browser: Safari

OS: Nexus 4 (Android 4.4.2)
Browser: Chrome Browser, Chrome Beta (latest version for both)

OS: Nexus 7 (2012 model) (Android 4.4.2)
Browser: Chrome Browser (latest version)

Hardware: LG G2 (LG-VS980)
OS: CyanogenMod 11 (nightly)
Android: 4.4.2
Chrome: 34.0.1847.114

I will continue to add to this list, but so far all mobile devices and all mobile browsers we've tested on.

**Comment by mourner** [comment]
This is super-weird. Do map examples on http://leafletjs.com/ have the same problem for you?

**Comment by chschie** [comment]
No, I don't see the issue when loading that url on mobile devices, nor do I see the problem when running your live examples.

The difference seems to be, in the example page on both, that the mapbox doesn't take up the full screen. The issue only appears to be an issue when running mapbox 'full screen' on mobile devices.

**Comment by mourner** [comment]
@chschie is there a problem with the map on the official mapbox.js examples page? https://www.mapbox.com/mapbox.js/api/v1.6.2/

**Comment by chschie** [comment]
Mourner,

No, no issue when running that example from any of the devices.

**Comment by chschie** [comment]
Here's a good example for you guys to try on any mobile device:

http://leafletjs.com/examples/quick-start-example.html

Simply tap the map repeatedly, and note that the message “You clicked the map…” appears for every other click, or incredibly inconsistently.

**Comment by mourner** [comment]
@chschie and if you zoom in the browser window enough, it doesn't, right?

**Comment by mourner** [comment]
Put this in your HTML and you should be fine:

``` html
<meta name="viewport" content="width=device-width, initial-scale=1.0">
```

**Comment by chschie** [comment]
We've already got our viewport setup for mobile (including the meta info you described above), so I don't believe that's a contributing factor.

**Comment by chschie** [comment]
"and if you zoom in the browser window enough, it doesn't, right?" 

I'm not sure how to interpret this statement?

**Comment by chschie** [comment]
I could still really use some help on this issue.

**Comment by mourner** [comment]
OK disregard that, but I still don't fully understand the conditions under which your issue is reproducible. Your statements contradict each other somewhat.

**Comment by chschie** [comment]
We're putting together a video of the issue, will post shortly.

**Comment by mourner** [comment]
@chschie I mean, could you set up a JSFiddle minimal test case where this issue is 100% reproducible? Or is this http://leafletjs.com/examples/quick-start-example.html page reflecting the issue fully?

**Comment by chschie** [comment]
http://youtu.be/pAy9jFdDFbY 

An example of leaflet 'missing' clicks on mobile.

**Comment by chschie** [comment]
We will have additional videos shortly showing other parts of the UI that displays the 'eating/missing' click issue. In the meantime, I'll see if I can do a jsfiddle.

**Comment by mourner** [comment]
In comments above you said that UI controls are affected, and this is a video of clicks on a map which is a different thing. Besides I see that in clicks that didn't work you dragged the map by a couple of pixels.

**Comment by chschie** [comment]
The UI controls are also affected, this seems to be showing up in quite a few places. I'll have another video up shortly, and full code on github for you to pull and run.

**Comment by chschie** [comment]
Vanilla mapbox install, showing this issue in the UI layer. Git Code up next.

http://www.youtube.com/watch?v=khR4hlGeKAU&feature=youtu.be

**Comment by chschie** [comment]
JS fiddle showing the exact problem, open on mobile, zoom in so can touch the zoom controls reliably, start pressing them, some of your clicks/touches will not register:

http://jsfiddle.net/kestes/2e5jQ/

Git code still coming.

**Comment by chschie** [comment]
Git Code Link:

https://github.com/kyleestes/mapbox-test

**Comment by chschie** [comment]
If there is anything else I can provide to help clarify the issue or help an investigation please let me know.

**Comment by mourner** [comment]
Try the latest Leaflet stable branch http://leaflet-cdn.s3.amazonaws.com/build/leaflet-stable.zip with Mapbox standalone.

**Comment by chschie** [comment]
I'd be happy to, but I'm not familiar with the steps to integrate leaflet into mapbox. Is there an updated mapbox.js file that has these changes implemented?

**Comment by kyleestes** [comment]
@mourner, thanks for your help with this issue. I wanted to let you know that not only does Leaflet seem to miss clicks on the map, but it also seems to miss clicks on controls such as the zoom in button. I was able to reproduce this on a desktop in Google Chrome on the tutorial titled Leaflet Quick Start Guide at http://leafletjs.com/examples/quick-start-example.html.

To illustrate the issue, I captured the following video: https://www.youtube.com/watch?v=5JG5M4vzAOU. In the video, I perform the following test:
1. Open the Leaflet Quick Start Guide in Google Chrome
2. Zoom in to level 18 in order to demonstrate that it is the maximum zoom level
3. Zoom out to level 0
4. Instrument the zoom in button to record and display the number of clicks in the Google Chrome development console
5. Click the zoom in button 18 times

Expected: the map is at zoom level 18
Actual: the map is not at zoom level 18

Speaking generally, it seems that Leaflet will miss exactly every other click when clicking rapidly (e.g., many times per second), and will miss no clicks when clicking slowly (e.g., once per second).

Lastly, it seems I can reproduce the above results even with the latest Leaflet stable branch you linked to above. Here is the JSFiddle I made in which I copied/pasted the contents of the leaflet-src.js file and the JavaScript code used in the Leaflet Quick Start Guide: http://jsfiddle.net/kestes/w7n7s/1/.

Please let me know if there is any other information I can provide to help identify the root cause of this issue.

**Comment by mourner** [comment]
@kyleestes the behavior that you describe is a completely different issue, in fact, it's expected behavior — Leaflet doesn't react to interactions while zoom animation is running. It was never considered a critical issue and no one brought this up earlier, so it certainly doesn't fit @chschie's initial report in which he says:

> it would be impossible to use this system in any kind of production environment on mobile

So I'm still pretty confused and not sure how to help you.

**Comment by lxbarth** [comment]
> I'd be happy to, but I'm not familiar with the steps to integrate leaflet into mapbox. Is there an updated mapbox.js file that has these changes implemented?

@chschie - you can always just use Leaflet directly without mapbox.js for a quick test.

@mourner - can you link to mapbox standalone with a quick instruction on how to combine it with Leaflet so @chschie can try the latest Leaflet stable?

**Comment by kyleestes** [comment]
Thanks @mourner and @lxbarth for the input. @chschie and I really appreciate your help. It is clear that clicks during zoom animations are intentionally ignored. I had thought that perhaps the issues were one in the same. Our next step is to perform additional testing with the Leaflet build that @mourner posted. Standby.

**Comment by mourner** [comment]
@kyleestes @chschie here's the Mapbox standalone link:

```
https://api.tiles.mapbox.com/mapbox.js/v1.6.2/mapbox.standalone.js
```

Just include it after including Leaflet (just JS; CSS can be just mapbox).

**Comment by kyleestes** [comment]
@mourner It appears that Leaflet behaves differently depending on the value of the user-agent HTTP request header. When the user-agent reflects a desktop browser such as Google Chrome for Windows, Leaflet misses zero clicks. But when the user-agent reflects a mobile browser such as Google Chrome for Andorid, Leaflet misses several clicks.

Here is an unlisted YouTube video that demonstrates Leaflet missing clicks with a mobile user-agent but not with a desktop user-agent (the clicks are performed programmatically via a script, so there is no map panning whatsoever): http://youtu.be/-nWlsivWoc0

In case you are curious, here is some documentation of the experiment I performed:
1.  On an Android mobile device, install the AutoTouch app
2.  On a desktop computer, download http://leaflet-cdn.s3.amazonaws.com/build/leaflet-stable.zip (the new Leaflet build posted by @mourner above)
3.  Extract `leaflet-stable.zip` to `C:\leaflet-stable`
4.  Create `C:\leaflet-stable\custom-icons-example.html` based on the custom icons example Leaflet tutorial from http://leafletjs.com/examples/custom-icons-example.html
   
   a. You can find a copy of `custom-icons-example.html` at https://github.com/kyleestes/leaflet-custom-icons-example/blob/master/custom-icons-example.html
5.  Connect the mobile device to the computer via USB and ensure Android debugging is enabled
6.  On a command prompt, run `adb push C:\leaflet-stable /storage/emulated/0/leaflet-stable`
7.  On the mobile device, run the `custom-icons-example.lua` AutoTouch script
   
   a. You can find a copy of `custom-icons-example.lua` at https://github.com/kyleestes/leaflet-custom-icons-example/blob/master/custom-icons-example.lua
8.  On the mobile device, open file:///storage/emulated/0/leaflet-stable/custom-icons-example.html
9.  On the mobile device, press volume down to start the AutoTouch script
10. Note that several clicks are missed
11. On the mobile device, re-run the `custom-icons-example.lua` AutoTouch script  
12. In the mobile browser, request the desktop site for file:///storage/emulated/0/leaflet-stable/custom-icons-example.html
13. On the mobile device, press volume down to start the AutoTouch script
14. Note that zero clicks are missed

For now, I plan to work around this issue by using a desktop user-agent. However, any other insight is much appreciated. Thanks again for all your help!

**Comment by tsfir** [comment]
Same Here

**Comment by tsfir** [comment]
I will be very happy if you somebody will help me understand how to work around this problem,
I develop a map module for my client for his new mobile  web site via Mapbox.
The development was running perfect until I discovered this issue (unfortunately only after finishing the map module) and in this condition the project will not be delivered to the client (at my testing the situation is worst than the videos something like 1:4 clicks is good click)

thanks in advance,

Ron

**Comment by chschie** [comment]
We ultimately changed the user-agent to make it run in desktop mode (where this isn't an issue). Risky, comes with unknowns, but in light of no better alternative, it appears to work.

**Comment by tsfir** [comment]
thanks for your response
I tried This code  with no luck - is this what you mean in changing the user-agent ?

```
<script>
    navigator.__defineGetter__('userAgent', function(){
       return 'Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/34.0.1847.137 Safari/537.36' 
    }); 
</script>
<script src='https://api.tiles.mapbox.com/mapbox.js/v1.6.3/mapbox.js'></script>
<link href='https://api.tiles.mapbox.com/mapbox.js/v1.6.3/mapbox.css' rel='stylesheet' />
```

**Comment by mourner** [comment]
You are probably hitting a workaround for an Android bug that made click fire two times on one tap. It puts a timer that ignores clicks that happen closer than 500ms to each other (the delay was 1000 but was reduced recently). Another lead to this is that the lua script above generates clicks 300ms apart.

Try removing these lines https://github.com/Leaflet/Leaflet/blob/master/src/dom/DomEvent.js#L77-L81, rebuilding Leaflet and seeing if problem went away. 

I still don't understand why this is a critical issue, since in my understanding it pretty much doesn't affect real user experience — no one clicks like mad in real world, it can only appear bad when stress-testing.

**Comment by tsfir** [comment]
Unfortunately its not  my case. I upload a "real user" video of my self clicking or "trying to click" on the map markers the first marker popup was opened after something like 5 clicks and the second I stoped counting.
http://1drv.ms/1ni5g0i

**Comment by mourner** [comment]
@tsfir your using a version that doesn't include some fixes on the matter. Please try using Leaflet stable branch build + Mapbox standalone.

**Comment by tsfir** [comment]
thanks @mourner 
Sorry on my low level but if you can please a little guide of what you mean "mapbox stand alone" how to use leaflet stable branch build will be very appreciated

**Comment by tsfir** [comment]
OK,  I think I figured it out,
did you meant by those ref ?

```
    <link rel="stylesheet" href="http://cdn.leafletjs.com/leaflet-0.7.2/leaflet.css" />
    <link href='//api.tiles.mapbox.com/mapbox.js/v1.6.3/mapbox.standalone.css' rel='stylesheet' />
    <script src="http://cdn.leafletjs.com/leaflet-0.7.2/leaflet.js"></script>
    <script src='http://api.tiles.mapbox.com/mapbox.js/v1.6.3/mapbox.standalone.js'></script>
```

and .... still the same- clicks are eaten up.

**Comment by mourner** [comment]
@tsfir yes, but try this Leaflet version instead (but don't use in production, it's for testing purposes): 

```
http://leaflet-cdn.s3.amazonaws.com/build/stable/leaflet.js
http://leaflet-cdn.s3.amazonaws.com/build/stable/leaflet.css
```

**Comment by tsfir** [comment]
Perfect !!!
Thanks you @mourner  so much... 
every click - works!
now what I am going to do with the production ?
I need to deliver this project in a month or so  (and I don't want to return to google API)

**Comment by mourner** [comment]
@tsfir Great to hear that! I've just released Leaflet 0.7.3 that includes the fix. I think mapbox.js minor release will follow shortly. :)

**Comment by tsfir** [comment]
Just in time (-:
Thanks Again

**Comment by techsin** [comment]
having same issue with markers, every other click seems to be completely ignored, but style attribute of html tag changes on every click regardless of popup opens or not. As if leaflet decides click wasn't good enough to open a popup. not sure what's happening.

more visible when you emulate mobile on chrome, and if you actually run it on android chrome it's barely works, like you have to tap 6 times to get one popup open and multiple times outside to close popup. Even you make icon 200x200 and very carefully touch to not cause drag it still doesn't open popup. 

Maybe i should just write my catch all event listener at map div and check if click was from marker

**Comment by techsin** [comment]
I also had same trouble but then I realized that somehow i was using 2013 version.

**Comment by mach1media** [comment]
This worked for me:

```
// create a DOM element for the marker
var el = document.createElement('div');

// give it a class name for styling purposes
el.className = 'c-gmap__marker';

// give it a unique ID
el.setAttribute("id", "marker_1");

// create the marker using the element just created
var marker_1 = new mapboxgl.Marker({element: el, color: '#404946'})
	.setPopup(
		new mapboxgl.Popup({anchor:'bottom'}).setHTML('...')
	)
	.setLngLat([XXX,YYY])
	.addTo(map);

// attach touch event to marker so it works on mobile
$('#marker_1').on('touchend', function(){ 
	
	// Check if there is already a popup on the map and if so, remove it
	$('.mapboxgl-popup').remove();
	
	// toggle its popup
	marker_1.togglePopup();
});
```
